In [27]:
import pandas as pd
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [14]:
df = pd.read_csv('/Users/damodargupta/Desktop/spam classification using naive bayes/spambase/spambase.data' )
df

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32            0.00              0.00                0.00   
1              0.14            0.28              0.21                0.07   
2              1.23            0.19              0.19                0.12   
3              0.63            0.00              0.31                0.63   
4              0.63            0.00              0.31                0.63   
...             ...             ...               ...                 ...   
4596           0.00            0.31              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.00            0.00              0.00                0.00   
4599           0.32            0.00              0.00                0.00   
4600           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0                0.00            0.00  ...        0.000        0.000   
1                0.00            0.94  ...        0.000        0.132   
2                0.64            0.25  ...        0.010        0.143   
3                0.31            0.63  ...        0.000        0.137   
4                0.31            0.63  ...        0.000        0.135   
...               ...             ...  ...          ...          ...   
4596             0.00            0.00  ...        0.000        0.232   
4597             0.00            0.00  ...        0.000        0.000   
4598             0.00            0.00  ...        0.102        0.718   
4599             0.00            0.00  ...        0.000        0.057   
4600             0.00            0.00  ...        0.000        0.000   

      char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0             0.0        0.778        0.000        0.000   
1             0.0        0.372        0.180        0.048   
2             0.0        0.276        0.184        0.010   
3             0.0        0.137        0.000        0.000   
4             0.0        0.135        0.000        0.000   
...           ...          ...          ...          ...   
4596          0.0        0.000        0.000        0.000   
4597          0.0        0.353        0.000        0.000   
4598          0.0        0.000        0.000        0.000   
4599          0.0        0.000        0.000        0.000   
4600          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          3.756                          61   
1                          5.114                         101   
2                          9.821                         485   
3                          3.537                          40   
4                          3.537                          40   
...                          ...                         ...   
4596                       1.142                           3   
4597                       1.555                           4   
4598                       1.404                      

In [15]:
df.isnull().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

In [17]:
# independent and dependent variables
X = df.iloc[: , :-1]
y = df.iloc[: , -1]

In [18]:
X

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32            0.00              0.00                0.00   
1              0.14            0.28              0.21                0.07   
2              1.23            0.19              0.19                0.12   
3              0.63            0.00              0.31                0.63   
4              0.63            0.00              0.31                0.63   
...             ...             ...               ...                 ...   
4596           0.00            0.31              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.00            0.00              0.00                0.00   
4599           0.32            0.00              0.00                0.00   
4600           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0                0.00            0.00  ...                   0.0        0.000   
1                0.00            0.94  ...                   0.0        0.000   
2                0.64            0.25  ...                   0.0        0.010   
3                0.31            0.63  ...                   0.0        0.000   
4                0.31            0.63  ...                   0.0        0.000   
...               ...             ...  ...                   ...          ...   
4596             0.00            0.00  ...                   0.0        0.000   
4597             0.00            0.00  ...                   0.0        0.000   
4598             0.00            0.00  ...                   0.0        0.102   
4599             0.00            0.00  ...                   0.0        0.000   
4600             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0           0.000          0.0        0.778        0.000        0.000   
1           0.132          0.0        0.372        0.180        0.048   
2           0.143          0.0        0.276        0.184        0.010   
3           0.137          0.0        0.137        0.000        0.000   
4           0.135          0.0        0.135        0.000        0.000   
...           ...          ...          ...          ...          ...   
4596        0.232          0.0        0.000        0.000        0.000   
4597        0.000          0.0        0.353        0.000        0.000   
4598        0.718          0.0        0.000        0.000        0.000   
4599        0.057          0.0        0.000        0.000        0.000   
4600        0.000          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          3.756                          61   
1                          5.114                         101   
2                          9.821                         485   
3                          3.537                          40   
4                          3.537              

In [19]:
y

0       1
1       1
2       1
3       1
4       1
       ..
4596    0
4597    0
4598    0
4599    0
4600    0
Name: spamornot, Length: 4601, dtype: int64

In [21]:
y.value_counts()

0    2788
1    1813
Name: spamornot, dtype: int64

# spam (1) or not (0)

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

# Bernoulli classifier

In [25]:
classifier = BernoulliNB()
classifier.fit(X_train , y_train)

BernoulliNB()

In [26]:
y_pred = classifier.predict(X_test)

In [28]:
print(classification_report(y_pred , y_test))
print(confusion_matrix(y_pred ,y_test))
print(accuracy_score(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.93      0.88      0.90       941
           1       0.82      0.89      0.85       578

    accuracy                           0.88      1519
   macro avg       0.87      0.88      0.88      1519
weighted avg       0.89      0.88      0.88      1519

[[824 117]
 [ 62 516]]
0.8821593153390388


# Multinomial classifier

In [29]:
classifier = MultinomialNB()
classifier.fit(X_train , y_train)

MultinomialNB()

In [30]:
y_pred = classifier.predict(X_test)

In [31]:
print(classification_report(y_pred , y_test))
print(confusion_matrix(y_pred ,y_test))
print(accuracy_score(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.85      0.80      0.82       948
           1       0.69      0.77      0.73       571

    accuracy                           0.79      1519
   macro avg       0.77      0.78      0.78      1519
weighted avg       0.79      0.79      0.79      1519

[[754 194]
 [132 439]]
0.7853851217906518


# Gaussian Classifier

In [32]:
classifier = GaussianNB()
classifier.fit(X_train , y_train)

GaussianNB()

In [33]:
y_pred = classifier.predict(X_test)

In [34]:
print(classification_report(y_pred , y_test))
print(confusion_matrix(y_pred ,y_test))
print(accuracy_score(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.73      0.95      0.83       681
           1       0.95      0.72      0.82       838

    accuracy                           0.82      1519
   macro avg       0.84      0.84      0.82      1519
weighted avg       0.85      0.82      0.82      1519

[[649  32]
 [237 601]]
0.8229098090849243


### in case of Bernoulli naive bayes classifier , we obtained an accuracy of 88.21 %

### in case of Multinomial naive bayes classifier , we obtained an accuracy of 78.53 %

### in case of Gaussian naive bayes classifier , we obtained an accuracy of 82.29 %

Thus, we get the best results in bernoulli classifier and we know that it works best in case of binary classification problems such as this problem where the two classes are the mail is a spam and it is not a spam